Jormun: Fix address in poi in api global autocomplete

kadhikari opened the pull request:
Ticket: https://jira.canaltp.fr/browse/NAVITIAII-2378

Commit:
jormungandr: admins added in poi.address

take poi coord in it's address in absent

Commit:
jomungandr: verify insee

Commit:
jomungandr: cleaning child objects depending on depth

kinnou02 commented:
Jenkins, test this please.

kinnou02 commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -127,12 +127,12 @@ def get_lon_lat(obj):
     coord = geocoding.get('coord', {})
-    lat = str(coord.get('lat')) if coord and coord.get('lat') else None
-    lon = str(coord.get('lon')) if coord and coord.get('lon') else None
+    lat = str(coord.get('lat')) if coord and coord.get('lat') else poi_lon
```
You mix lat and lon here

kinnou02 commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -217,7 +218,7 @@ def output(self, key, obj):
                 create_administrative_regions_field(geocoding) or create_admin_field(geocoding),
             "properties": {p.get('key'): p.get('value') for p in geocoding.get("properties", [])},
-            "address": create_address_field(geocoding.get("address"))
+            "address": create_address_field(geocoding.get("address"), lat, lon)
```
I might be paranoiac, but I will prefer that we use named parameter for lat and lon, as: 
```
create_address_field(geocoding.get("address"), lat=lat, lon=lon)
```

kinnou02 commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -349,10 +350,47 @@ def _check_response(cls, response, uri):
+    def _clean_response(cls, response, depth=3):
+
+        def _clear_object(obj, dep):
+            if dep == -1:
```
Since we are already checking the `depth` before calling this method we don't have to do it another time

kinnou02 commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -349,10 +350,47 @@ def _check_response(cls, response, uri):
+
+            if _depth == -1:
+                _clear_object(_value, _depth)
+            elif _key == 'administrative_regions':
```
Why? a comment might help

kinnou02 commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -349,10 +350,47 @@ def _check_response(cls, response, uri):
+        if features:
+            for feature in features:
+                key = 'geocoding'
+                value = feature.get('properties').get('geocoding')
```
`properties` or `geocoding` might be None

Commit:
jormungandr: some corrections on comments

kadhikari commented:
@kinnou02 could you have a look please?

pbougue commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -127,12 +127,12 @@ def get_lon_lat(obj):
 
 
-def create_address_field(geocoding):
+def create_address_field(geocoding, poi_lat, poi_lon):
```
Why not populate `geocoding.address.coord`before (if empty) instead of adding params?
If we keep those params, default them to None might be a bit better.

kadhikari replied:
I think we parse and modify two times if i modify the code as you suggest?

pbougue commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -418,8 +460,9 @@ def get(self, request, instances):
             method = requests.post
 
         raw_response = self.call_bragi(url, method, **kwargs)
+        depth = request.get('depth', 3)
```
Default depth is 1 (so I guess this should be the default value everywhere)

pbougue commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -349,10 +350,49 @@ def _check_response(cls, response, uri):
+        return response
+
+    @classmethod
+    def response_marshaler(cls, response_bragi, uri=None, depth=3):
```
depth=1 by default

pbougue commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -349,10 +350,49 @@ def _check_response(cls, response, uri):
 
     @classmethod
-    def response_marshaler(cls, response_bragi, uri=None):
+    def _clean_response(cls, response, depth=3):
```
depth=1 by default

pbougue commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -349,10 +350,49 @@ def _check_response(cls, response, uri):
-    def response_marshaler(cls, response_bragi, uri=None):
+    def _clean_response(cls, response, depth=3):
+
+        def _clear_object(obj):
```
Minor: might even be in utils

pbougue commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -349,10 +350,49 @@ def _check_response(cls, response, uri):
+            elif isinstance(obj, dict):
+                obj.clear()
+
+        def _manage_depth(_key, _value, _depth):
```
This feels too generic, more than it is in kraken...
This code feels to clean :stuck_out_tongue_closed_eyes: 
Let's just compare results obtained in jormun using bragi vs. those using kraken (that's what you're supposed to do in tests)

TeXitoi commented on source/jormungandr/jormungandr/interfaces/v1/serializer/geocode_json.py:
```diff
@@ -169,7 +169,8 @@ def get_address(self, obj):
             return None
-        return create_address_field(address)
+        poi_lon, poi_lat = get_lon_lat(obj)
+        return create_address_field(address, poi_lat, poi_lon)
```
Same here, something like `return create_address_field(address, lat=poi_lat, lon=poi_lon)`?

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -279,7 +399,8 @@ def test_autocomplete_call(self):
             'limit': 10,
-            'pt_dataset': 'main_routing_test'
+            'pt_dataset': 'main_routing_test',
+            'depth': 1
```
This should not be necessary as it's default value

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -323,7 +444,8 @@ def test_autocomplete_call_override(self):
             'limit': 10,
-            'pt_dataset': 'main_routing_test'
+            'pt_dataset': 'main_routing_test',
+            'depth': 1
```
This should not be necessary as it's default value

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -509,6 +631,149 @@ def test_stop_area_without_modes(self):
+            'type[]': [u'public_transport:stop_area', u'street', u'house', u'poi', u'city'],
+            'limit': 10,
+            'pt_dataset': 'main_routing_test',
+            'depth': 1
```
Nice to fix it in one test, though (please let that one ;) )

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -674,131 +940,20 @@ def test_journey_with_external_uri_from_bragi(self):
             u'pt_dataset': 'main_routing_test',
             u'type[]': [u'public_transport:stop_area', u'street', u'house', u'poi', u'city'],
             u'limit': 10,
+            u'depth': 1,
```
This should not be necessary as it's default value

kadhikari replied:
This parameter is needed for the Mock.

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -825,11 +980,13 @@ def get_autocomplete(query):
                 assert response_from['poi']['properties']["capacity"] == "20"
                 assert response_from['poi']['properties']["ref"] == "12"
-                assert response_from['poi']['address']['id'] == "179077655"
+                assert response_from['poi']['address']['id'] == "addr:8.9809147;42.561667"
```
This is not a correct address id: you don't get a prefix from kraken autocomplete
This was also part of the JIRA comment (the minor part of the 3)

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -509,6 +631,149 @@ def test_stop_area_without_modes(self):
+            address = poi['address']
+            assert address['coord']['lat'] == '8.98312e-05'
+            assert address['coord']['lon'] == '8.98312e-05'
+            assert address['id'] == "addr:8.9809147;42.561667"
```
wrong id

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -509,6 +631,149 @@ def test_stop_area_without_modes(self):
+            address = poi['address']
+            assert address['coord']['lat'] == '8.98312e-05'
+            assert address['coord']['lon'] == '8.98312e-05'
+            assert address['id'] == "addr:8.9809147;42.561667"
```
wrong id

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -509,6 +631,149 @@ def test_stop_area_without_modes(self):
+            address = poi['address']
+            assert address['coord']['lat'] == '8.98312e-05'
+            assert address['coord']['lon'] == '8.98312e-05'
+            assert address['id'] == "addr:8.9809147;42.561667"
```
wrong id

Commit:
jormungandr: modifications after remarks

Results of queries with depth= 0, 1, 2 and 3 compared between kraken and
bragi.

Commit:
jormungandr: minor correction

kadhikari commented:
@pbougue i pushed a version with a complete verification of queries with depth = 0, 1, 2 and 3

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -509,6 +633,183 @@ def test_stop_area_without_modes(self):
+            assert poi_admins[0]['coord']['lon'] == "2.3957517"
+
+            address = poi['address']
+            assert address['coord']['lat'] == '8.98312e-05'
```
Please have different values in lat/lon so we know if we mixed them (same everywhere)

pbougue commented on source/jormungandr/tests/bragi_autocomplete_tests.py:
```diff
@@ -509,6 +633,183 @@ def test_stop_area_without_modes(self):
+            address = poi['address']
+            assert address['coord']['lat'] == '8.98312e-05'
+            assert address['coord']['lon'] == '8.98312e-05'
+            assert address['id'] == "8.98312e-05;8.98312e-05"
```
same: lat != lon

pbougue reviewed (commented):
This also has to be hand-tested (I can give you my conf)

Commit:
jormungandr: admin name without insee and some other modifications

Commit:
jormungandr: admin.name added in mock bragi

pbougue commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -93,7 +93,7 @@ def create_administrative_regions_field(geocoding):
         response.append({
             "insee": admin.get('insee'),
-            "name": admin.get('label'),
+            "name": admin.get('name'),
```
:+1:

pbougue commented on source/jormungandr/jormungandr/interfaces/v1/test/elastic_search_tests.py:
```diff
@@ -63,6 +63,7 @@ def bragi_house_jaures_feature():
                         "insee": "02000",
                         "level": 8,
                         "label": "Saint-Quentin",
```
It would be nicer to have realistic labels on the ones tested (on all of them would be even nicer).
If not realistic, at least different from name, as mixing them was a bug before.

pbougue commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -349,10 +357,53 @@ def _check_response(cls, response, uri):
+                _clear_object(_value)
+            elif isinstance(_value, dict):
+                for k, v in _value.items():
+                    if is_deleteable(k, v, _depth):
```
Isn't that part already managed 2 lines below (as you call it with depth-1) ?

pbougue reviewed (approved):
This looks good.
What worries me is that the cleaning impacts every objects, not just poi.address.administrative_regions...So we need to test it for all depth, and all other features (stops, address, admin, streets), at least manually.

benoit-bst commented on source/jormungandr/jormungandr/autocomplete/geocodejson.py:
```diff
@@ -127,15 +127,15 @@ def get_lon_lat(obj):
+    lon = str(coord.get('lon')) if coord and coord.get('lon') else poi_lon
+    address_id = '{lon};{lat}'.format(lon=lon, lat=lat)
+    resp = {
+        "id": address_id,
```
Why the Poi address ID is now long;lat ? It doesn't exist an associated ID ?

TeXitoi replied:
address ids are coords. If we don't have reverse geocoding, an empty address can be generated with such an id.

benoit-bst replied:
:+1:

Commit:
jormungandr: modifications not to  display empty element

Many tests added
